Feature: Add Pinch (in/out) support

pixel-shock opened the pull request:
@sds100 As I said, I added Pinch support

What I did:

- add pinch support
- fix max finger count on swipe and pinch gestures (seems so be a bug in Android that you're not able to use 10 fingers, because in the code of the GestureDescription.java:176 the devs wrote "if (mStrokes.size() >= MAX_STROKE_COUNT)" and "MAX_STROKE_COUNT" is 10 so 9 fingers are the maximum currently (added this to the label of the input fields and to the validator for the save action button enabled)
- change icons for swipe and pinch gestures
- solved some code format issues (redundant semicolons etc)

sds100 commented:
I'll take a look at this PR later this week. I don't understand the issue with the max 10 fingers bug. Doesn't `if (mStrokes.size() >= MAX_STROKE_COUNT)` allow 10 fingers?

pixel-shock commented:
No hurries 😉 

No, the ">=" includes 10, which is really strange because normally everyone should have max 10 fingers. If I allow 10 fingers in the input fields then the GestureBuilder will throw an Error with "too many strokes" because of that line ...

The line `if (mStrokes.size() >= MAX_STROKE_COUNT)` is the if-clause for the exception, so if the strokes are more than 9, it will throw the error ... really strange in my opinion

sds100 commented:
Ah, I see now. Thats the catch for the exception. Yeah, thats strange.

pixel-shock commented:
Jep, really strange, but maybe it would be helpful to push the min required API level of the app and then doing the check like `GestureDescription.getMaxStrokeCount()` to be sure that the finger count always matches the requirements? The touch gestures will be triggered at min level 24 anyways right? So it would make sense to push the min API level **BUT** if you don't like this I can remove my last commit

pixel-shock commented:
I've added a check for StrokeDuration and "finger"Count with a proper message ... and reverted my last 2 commits regarding a min sdk version bump, so if the user enters a duration which is too high or setup a finger count which is too high, then it will trigger a proper error message

sds100 commented:
Hi @pixel-shock, can you also document this feature in the `actions.md` file in the docs folder? We have a website documenting features in the app. Please put any limitations of the action there as well.

pixel-shock commented:
Hi,


sure, I'll do it tomorrow 👍😉

pixel-shock commented:
@sds100 Added the docs, also for the swipe gesture. If I missed something please let me know 👍

sds100 commented:
Can you explain what Radius is in the UI? Maybe suggest good values. It took me a while to figure it out. It is how far away from the "end" coordinates the fingers should be right?

pixel-shock commented:
Hi,

jep exactly. Maybe it should be also "distance" to be more clear? Maybe radius is a little bit to much math-based?

sds100 commented:
yeah, "pinch distance" or something?

pixel-shock commented:
I was to quick, I've just added more explanation to the docs, but I'll change the "radius" to "Pinch distance", makes more sense 👍

pixel-shock commented:
Changed it to "Pinch distance" and changed/added the docs in the actions markdown file 😉

sds100 commented on app/src/main/java/io/github/sds100/keymapper/util/MathUtils.kt:
```diff
@@ -52,3 +54,21 @@ fun movePointByDistanceAndAngle(p: Point, distance: Int, degrees: Double): Point
     return rad2deg(atan2((p2.y - p1.y).toDouble(), (p2.x - p1.x).toDouble()))
 }
+
+fun distributePointsOnCircle(circleCenter: Point, circleRadius: Float, numPoints: Int): ArrayList<Point> {
```
Use `List` instead of ArrayList in Kotlin.

pixel-shock replied:
fixed

sds100 commented on app/src/main/java/io/github/sds100/keymapper/util/ui/ResourceProvider.kt:
```diff
@@ -65,4 +69,5 @@ interface ResourceProvider {
     fun getText(@StringRes resId: Int): CharSequence
     fun getDrawable(@DrawableRes resId: Int): Drawable
     fun getColor(@ColorRes color: Int): Int
+    fun getContext(): Context;
```
This is dangerous! We don't want to be exposing Context in viewmodels. It is a code-smell and can also lead to memory leaks. The SpinnerAdapter should be in the Fragment since it is only needed for the View.

pixel-shock replied:
Sorry about that. I knew you will mention it, but as a lag of knowhow in Android Dev this was my only solution. I changed it and removed the getContext function

pixel-shock commented:
Morning @sds100,

I changed "radius" to "distance" in code too, so the code will match the UI to avoid confusions. And as we're talking about "distance" I also changed the param in the "distributePointsOnCircle" function to be divided by 2 because otherwise the distance would be like 200 but as the function uses a radius to distribute the points the gesture size would be 400 then, you know what I mean?

sds100 commented:
Yeah, that makes sense. I'll look over it again today. Thanks for the work!

pixel-shock commented:
No worries, I am happy to get a review on that one because I have **a lot to learn** and as I am working on the next feature for interacting with elements on the screen it will help me a lot 😉

sds100 commented on app/src/main/java/io/github/sds100/keymapper/actions/pinchscreen/PichPickDisplayCoordinateFragment.kt:
```diff
@@ -0,0 +1,153 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        pinchTypesDisplayValues = arrayOf(getString(io.github.sds100.keymapper.R.string.hint_coordinate_type_PINCH_IN), getString(io.github.sds100.keymapper.R.string.hint_coordinate_type_PINCH_OUT)).toMutableList();
```
You should be able to import the `R` in `R.string.hint_coordinate_type_PINCH_IN` without using the absolute package name.

pixel-shock replied:
Sure, that was a strange thing in Android Studio ... it won't let me use it until I manually added the proper import for that ...

sds100 commented on app/src/main/java/io/github/sds100/keymapper/actions/pinchscreen/PichPickDisplayCoordinateFragment.kt:
```diff
@@ -0,0 +1,153 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        pinchTypesDisplayValues = arrayOf(getString(io.github.sds100.keymapper.R.string.hint_coordinate_type_PINCH_IN), getString(io.github.sds100.keymapper.R.string.hint_coordinate_type_PINCH_OUT)).toMutableList();
```
Also, for consistency reasons can you keep the string resource names lower case

pixel-shock replied:
fixed 👍

pixel-shock commented on app/src/main/res/values/strings.xml:
```diff
@@ -152,10 +157,14 @@
+    <string name="hint_coordinate_y_start">Start Y</string>
+    <string name="hint_coordinate_x_end">End X</string>
+    <string name="hint_coordinate_y_end">End Y</string>
+    <string name="hint_coordinate_radius">Pinch distance</string>
```
fixed 👍

pixel-shock commented on app/src/main/java/io/github/sds100/keymapper/system/accessibility/MyAccessibilityService.kt:
```diff
@@ -465,6 +468,61 @@ class MyAccessibilityService : AccessibilityService(), LifecycleOwner, IAccessib
+
+            val gestureBuilder = GestureDescription.Builder()
+
+            var startCoordinates = arrayListOf<Point>()
```
I change the code to only use 1 NON mutable list for the dynamic created coordinates, the other list is not really necessary, my fault.

pixel-shock commented:
thanks for the review Seth, really appreciate that 👍

sds100 commented:
Also, in the future can you fork from `develop` instead of master.

pixel-shock commented:
Superb. Thank you very much 🍻

pixel-shock commented:
Hmmm ... but I think the current merge conflict is caused by the diverged deleop/master right? The Swipe Screen Gesture is missing in develop currently

sds100 commented:
Oh right, I'll fix that merge conflict. Also, can I set 200ms as a good default for the duration? I'm also working on showing an error if the finger count is less than 2.

sds100 commented:
is a duration of 0 allowed or not?

pixel-shock commented:
Sure, duration can have a default of 200ms ... 0 should not allowed. I see my fault in the "isOptionsValid" function, sorry for that ... it should be duration > 0 

For the fingercount: It should not be possible to enter less than 2 fingers

sds100 commented:
I'm fixing those issues right now pinching. I'll let you do the work of copying what I did for the old swipe action as well. I would like more error messages there as well.

pixel-shock commented:
alright 👍

sds100 commented:
10 is the maximum allowed finger count?

pixel-shock commented:
depends on the Android Version, on Nougat it is 9 (see our discussion about that), that's why I did the check in the accessiblity service, I think they fixed that on newer releases

pixel-shock commented:
```
if (fingerCount >= GestureDescription.getMaxStrokeCount()) {
    return Error.GestureStrokeCountTooHigh
}
if (duration >= GestureDescription.getMaxGestureDuration()) {
    return Error.GestureDurationTooHigh
}
```

sds100 commented:
yup, just trying to create an error message for it. it is specific to SDK 24 the issue with it being 9?

pixel-shock commented:
Did not checked that to be honest. I would suggest to add an error like: 

The entered Number does not match the current Android limitstion of X

That would be very generic

sds100 commented:
I'll just say its 10. does the app crash on Nougat if you try to use 10?

sds100 commented:
I just checked and Nougat is 0.7% of active users currently 😄

pixel-shock commented:
No it will just throw an error via Toast Notification

sds100 commented:
I also noticed that you didn't put the functions inside `MathUtils` inside a `MathUtils` object. I don't like having functions in the global namespace because its unclear where a functions comes from. I've fixed this but just so you know for next time 🙂

pixel-shock commented:
Good point. Thank you very much! 🍻

As I have a cold I will have a look into everything tomorrow. For today I just need some sleep 😬🤷‍♂️

sds100 commented:
Oh no! Hopefully it doesn't last long 😄 I'll try to be complete with my review tonight

sds100 commented on app/src/main/java/io/github/sds100/keymapper/actions/pinchscreen/PinchPickDisplayCoordinateViewModel.kt:
```diff
@@ -0,0 +1,240 @@
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val fingerCountError: StateFlow<String?> = fingerCount.map { count ->
```
can you do the same for `SwipePickDisplayCoordinateViewModel`. Check the xml file for how to bind this with the view

pixel-shock replied:
done 👍

sds100 commented on app/src/main/java/io/github/sds100/keymapper/actions/pinchscreen/PinchPickDisplayCoordinateViewModel.kt:
```diff
@@ -0,0 +1,240 @@
+        it.toString()
+    }.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    val durationError: StateFlow<String?> = duration.map { d ->
```
can you do the same for `SwipePickDisplayCoordinateViewModel`

pixel-shock replied:
done 👍

pixel-shock commented:
Thx. I Hope that too, I am going into vacation next weekend. Tuesday goLive with a project at work so lot to do 🤪

Awesome. Thx again 🙌 I will try to fix everything then tomorrow 👍

pixel-shock commented:
fyi: I could not resist to add androids maxstrokes to the fingercount error - sorry 🤣

sds100 reviewed (approved):
looks good!

DagJenssen commented:
I am sorry if I ask in the wrong place, but since the thread is open for comments I will take my chances:
I am setting up Key Mapper to allow a six button bluetooth remote controller which sends keyboard strokes to the Android device to zoom, pan and go back to position in Google Maps.
It works very well with the swipe feature, but pinch seems to only pinch one way. Whether I choose pinch out or in, it will zoom out. Is this a bug or have I misunderstood the functionality?
Coordinates entered are in the middle of the screen and pinch distance is the default, 200px in 200ms.
I would be very happy for some assistance on a great app.
Dag
